File: sdk/python/velocity_sdk/testing.py
"""
VELOCITY-WorkFlow Python SDK - Testing utilities.

Provides test environment and mock client for unit testing workflows
without requiring a running VELOCITY-WorkFlow server.
"""

from typing import Any, Dict, List, Optional
from .client import VelocityClient, WorkflowHandle, WorkflowDescription, WorkflowStatus
from .exceptions import WorkflowNotFoundError, WorkflowAlreadyCompletedError


class MockVelocityClient:
    """Mock client for testing workflows without a server."""

    def __init__(self):
        self._workflows: Dict[int, Dict[str, Any]] = {}
        self._signals: Dict[int, List[Dict[str, Any]]] = {}
        self._next_key = 1

    def start_workflow(
        self,
        workflow_type: str,
        namespace: str = "default",
        task_queue: str = "default",
        total_steps: int = 1,
        input_data: Optional[bytes] = None,
    ) -> WorkflowHandle:
        """Start a mock workflow."""
        key = self._next_key
        self._next_key += 1

        self._workflows[key] = {
            "workflow_type": workflow_type,
            "namespace": namespace,
            "task_queue": task_queue,
            "total_steps": total_steps,
            "current_step": 0,
            "status": WorkflowStatus.RUNNING,
            "result": None,
        }
        self._signals[key] = []

        return WorkflowHandle(workflow_key=key, workflow_id=key, run_id=key + 1000)

    def describe_workflow(self, workflow_key: int) -> WorkflowDescription:
        """Describe a mock workflow."""
        if workflow_key not in self._workflows:
            raise WorkflowNotFoundError(workflow_key)

        wf = self._workflows[workflow_key]
        return WorkflowDescription(
            workflow_key=workflow_key,
            status=wf["status"],
            current_step=wf["current_step"],
            total_steps=wf["total_steps"],
        )

    def signal_workflow(
        self,
        workflow_key: int,
        signal_name: str,
        payload: bytes = b"",
    ) -> bool:
        """Send a signal to a mock workflow."""
        if workflow_key not in self._workflows:
            raise WorkflowNotFoundError(workflow_key)

        self._signals[workflow_key].append({
            "signal_name": signal_name,
            "payload": payload,
        })
        return True

    def complete_workflow(self, workflow_key: int, result: bytes = b"") -> bool:
        """Complete a mock workflow."""
        if workflow_key not in self._workflows:
            raise WorkflowNotFoundError(workflow_key)

        wf = self._workflows[workflow_key]
        if wf["status"] != WorkflowStatus.RUNNING:
            raise WorkflowAlreadyCompletedError(workflow_key)

        wf["status"] = WorkflowStatus.COMPLETED
        wf["result"] = result
        return True

    def fail_workflow(self, workflow_key: int, reason: str = "") -> bool:
        """Fail a mock workflow."""
        if workflow_key not in self._workflows:
            raise WorkflowNotFoundError(workflow_key)

        wf = self._workflows[workflow_key]
        if wf["status"] != WorkflowStatus.RUNNING:
            raise WorkflowAlreadyCompletedError(workflow_key)

        wf["status"] = WorkflowStatus.FAILED
        return True

    def cancel_workflow(self, workflow_key: int) -> bool:
        """Cancel a mock workflow."""
        if workflow_key not in self._workflows:
            raise WorkflowNotFoundError(workflow_key)

        self._workflows[workflow_key]["status"] = WorkflowStatus.CANCELED
        return True

    def get_signals(self, workflow_key: int) -> List[Dict[str, Any]]:
        """Get all signals received by a workflow."""
        return self._signals.get(workflow_key, [])

    def close(self):
        """No-op for mock client."""
        pass


class WorkflowTestEnvironment:
    """Test environment for running workflows in isolation."""

    def __init__(self):
        self.client = MockVelocityClient()
        self._time_offset = 0

    def start_workflow(
        self,
        workflow_type: str,
        **kwargs,
    ) -> WorkflowHandle:
        """Start a workflow in the test environment."""
        return self.client.start_workflow(workflow_type, **kwargs)

    def complete_workflow(self, workflow_key: int, result: bytes = b"") -> bool:
        """Complete a workflow in the test environment."""
        return self.client.complete_workflow(workflow_key, result)

    def signal_workflow(
        self,
        workflow_key: int,
        signal_name: str,
        payload: bytes = b"",
    ) -> bool:
        """Signal a workflow in the test environment."""
        return self.client.signal_workflow(workflow_key, signal_name, payload)

    def time_skip(self, seconds: int) -> None:
        """Advance the test environment's clock."""
        self._time_offset += seconds

    def get_current_time(self) -> int:
        """Get the current test time (real time + offset)."""
        import time
        return int(time.time()) + self._time_offset

    def assert_workflow_completed(self, workflow_key: int) -> None:
        """Assert that a workflow has completed."""
        desc = self.client.describe_workflow(workflow_key)
        assert desc.status == WorkflowStatus.COMPLETED, (
            f"Expected workflow {workflow_key} to be completed, but status is {desc.status}"
        )

    def assert_signal_received(self, workflow_key: int, signal_name: str) -> None:
        """Assert that a workflow received a specific signal."""
        signals = self.client.get_signals(workflow_key)
        signal_names = [s["signal_name"] for s in signals]
        assert signal_name in signal_names, (
            f"Expected signal '{signal_name}' not found. Received: {signal_names}"
        )

    def reset(self) -> None:
        """Reset the test environment."""
        self.client = MockVelocityClient()
        self._time_offset = 0


def assert_workflow_completed(client: MockVelocityClient, workflow_key: int) -> None:
    """Assert that a workflow has completed."""
    desc = client.describe_workflow(workflow_key)
    assert desc.status == WorkflowStatus.COMPLETED, (
        f"Expected workflow {workflow_key} to be completed, but status is {desc.status}"
    )


def assert_signal_received(
    client: MockVelocityClient,
    workflow_key: int,
    signal_name: str,
) -> None:
    """Assert that a workflow received a specific signal."""
    signals = client.get_signals(workflow_key)
    signal_names = [s["signal_name"] for s in signals]
    assert signal_name in signal_names, (
        f"Expected signal '{signal_name}' not found. Received: {signal_names}"
    )

File: sdk/ruby/lib/velocity_sdk/testing.rb
# frozen_string_literal: true

module VelocitySdk
  # ─── Testing utilities ───────────────────────────────────────────────────────
  #
  # Provides TestWorkflowEnvironment and MockClient for unit-testing
  # workflow logic without a running engine or server.

  # In-memory mock client that mirrors the VelocityClient API surface.
  #
  # Useful for unit tests that need to verify workflow interactions without
  # depending on the real engine or FFI library.
  class MockClient
    # @return [Hash<Integer, Hash>] Internal workflow state store.
    attr_reader :workflows

    # @return [Hash<Integer, Array<Hash>>] Signal store per workflow.
    attr_reader :signals

    def initialize
      @workflows = {}
      @signals = {}
      @next_key = 1
    end

    # Start a mock workflow.
    # @return [Integer] Workflow key.
    def start_workflow(workflow_type, namespace: 'default', task_queue: 'default', total_steps: 1, input: '')
      key = @next_key
      @next_key += 1

      @workflows[key] = {
        workflow_type: workflow_type,
        namespace: namespace,
        task_queue: task_queue,
        total_steps: total_steps,
        current_step: 0,
        status: WorkflowStatus::RUNNING,
        result: nil,
      }
      @signals[key] = []
      key
    end

    # Describe a mock workflow.
    # @return [Hash] Workflow description.
    # @raise [WorkflowNotFoundError]
    def describe_workflow(workflow_key)
      wf = @workflows[workflow_key]
      raise WorkflowNotFoundError.new(workflow_key) unless wf

      {
        workflow_key: workflow_key,
        status: wf[:status],
        current_step: wf[:current_step],
        total_steps: wf[:total_steps],
        namespace: wf[:namespace],
        result: wf[:result],
      }
    end

    # Signal a mock workflow.
    # @raise [WorkflowNotFoundError]
    def signal_workflow(workflow_key, signal_name, payload = '')
      raise WorkflowNotFoundError.new(workflow_key) unless @workflows.key?(workflow_key)

      @signals[workflow_key] << { signal_name: signal_name, payload: payload }
      true
    end

    # Complete a mock workflow.
    # @raise [WorkflowNotFoundError]
    # @raise [WorkflowAlreadyCompletedError]
    def complete_workflow(workflow_key, result = '')
      wf = @workflows[workflow_key]
      raise WorkflowNotFoundError.new(workflow_key) unless wf
      raise WorkflowAlreadyCompletedError.new(workflow_key) unless wf[:status] == WorkflowStatus::RUNNING

      wf[:status] = WorkflowStatus::COMPLETED
      wf[:result] = result
      true
    end

    # Fail a mock workflow.
    # @raise [WorkflowNotFoundError]
    # @raise [WorkflowAlreadyCompletedError]
    def fail_workflow(workflow_key, reason = '')
      wf = @workflows[workflow_key]
      raise WorkflowNotFoundError.new(workflow_key) unless wf
      raise WorkflowAlreadyCompletedError.new(workflow_key) unless wf[:status] == WorkflowStatus::RUNNING

      wf[:status] = WorkflowStatus::FAILED
      true
    end

    # Cancel a mock workflow.
    # @raise [WorkflowNotFoundError]
    def cancel_workflow(workflow_key)
      wf = @workflows[workflow_key]
      raise WorkflowNotFoundError.new(workflow_key) unless wf

      wf[:status] = WorkflowStatus::CANCELED
      true
    end

    # Get the status of a mock workflow.
    # @return [Integer] Status code.
    # @raise [WorkflowNotFoundError]
    def get_status(workflow_key)
      wf = @workflows[workflow_key]
      raise WorkflowNotFoundError.new(workflow_key) unless wf

      wf[:status]
    end

    # Get all signals received by a workflow.
    # @return [Array<Hash>]
    def get_signals(workflow_key)
      @signals[workflow_key] || []
    end

    # List all workflow keys.
    # @return [Array<Integer>]
    def list_workflows
      @workflows.keys
    end

    # Close the mock client (no-op).
    def close; end
  end

  # Isolated test environment wrapping a MockClient.
  #
  # Provides assertion helpers and time-skip support for deterministic tests.
  class TestWorkflowEnvironment
    # @return [MockClient] The underlying mock client.
    attr_reader :client

    def initialize
      @client = MockClient.new
      @time_offset_secs = 0
    end

    # Start a workflow in the test environment.
    # @return [Integer] Workflow key.
    def start_workflow(workflow_type, **kwargs)
      @client.start_workflow(workflow_type, **kwargs)
    end

    # Complete a workflow in the test environment.
    def complete_workflow(workflow_key, result = '')
      @client.complete_workflow(workflow_key, result)
    end

    # Signal a workflow in the test environment.
    def signal_workflow(workflow_key, signal_name, payload = '')
      @client.signal_workflow(workflow_key, signal_name, payload)
    end

    # Advance the simulated clock.
    # @param seconds [Integer]
    def time_skip(seconds)
      @time_offset_secs += seconds
    end

    # Current simulated time as UNIX epoch seconds.
    # @return [Integer]
    def current_time_secs
      Time.now.to_i + @time_offset_secs
    end

    # Assert that a workflow has completed.
    # @raise [RuntimeError] If the workflow is not completed.
    def assert_workflow_completed(workflow_key)
      desc = @client.describe_workflow(workflow_key)
      return if desc[:status] == WorkflowStatus::COMPLETED

      raise "Expected workflow #{workflow_key} to be completed, but status is #{WorkflowStatus.name_for(desc[:status])}"
    end

    # Assert that a workflow received a specific signal.
    # @raise [RuntimeError] If the signal was not received.
    def assert_signal_received(workflow_key, signal_name)
      names = @client.get_signals(workflow_key).map { |s| s[:signal_name] }
      return if names.include?(signal_name)

      raise "Expected signal '#{signal_name}' not found for workflow #{workflow_key}. Received: #{names.join(', ')}"
    end

    # Reset the environment to a clean state.
    def reset
      @client = MockClient.new
      @time_offset_secs = 0
    end
  end
end
